perf(highlight): allow decoration providers to skip ranges without data

bfredl opened the pull request:
Continuing the work of #31400

That PR allowed the provider to be invoked multiple times per line. But we only want to do that when there actually is more data later on the line, otherwise it is just lua VM overhead for no good. Additionally, we want to skip over lines which contain no new highlight items. The TS query cursor already tells us what the next position with more data is, so there is no need to reinvoke the range callback before that.

NB: this removes the double buffering introduced in #32619 which is fundamentally incompatible with this (nvim core is supposed to keep track of long ranges by itself, without requiring a callback reinvoke blitz). Need to adjust the priorities some other way to fix the same issue.

<!--
  Thank you for contributing to Neovim!
  If this is your first time, check out https://github.com/neovim/neovim/blob/master/CONTRIBUTING.md#pull-requests-prs
  for our PR guidelines.
-->

bfredl committed:
perf(highlight): allow decoration providers to skip ranges without data

Continuing the work of #31400

That PR allowed the provider to be invoked multiple times per line.
We want only to do that when there actually is more data later on the
line. Additionally, we want to skip over lines which contain no new
highlight items. The TS query cursor already tells us what the next
position with more data is, so there is no need to reinvoke the range
callback before that.

NB: this removes the double buffering introduced in #32619 which
is funtamentally incompatible with this (nvim core is supposed to keep
track of long ranges by itself, without requiring a callback reinvoke
blitz). Need to adjust the priorities some other way to fix the same issue.

bfredl commented:
a solution for the original problem might actually be to just adjust the priorities in `highlight.scm` without any extra mechanism. 

The "priority" property exists in order to be used, which it was not in the example query. if it is to be expected that `comment` contains general elements of the same language which should take priority over interleaving comment continuation chars, then priority of comment should be lowered.

clason commented:
General note: Templating languages (which is where this issue comes from) are simply cursed and in general a poor fit for tree-sitter's hierarchical injection model. I have often said that these should be handled by dedicated plugins that ship a bespoke set of queries. (Of course, this would require buffer-local queries, which is a general and high-level desideratum.)

vanaigr commented:
There are some cases where the same node can be both a container and be contained within other nodes, in which case the tree depth would determine which highlighting should be applied, e.g.:

```javascript
element.innerHTML = '<img onclick="/* I am a comment inside of 2 strings */">'
```

Which nvim-treesitter currently highlights as expected:

<img width="1105" height="49" alt="image" src="https://github.com/user-attachments/assets/216544a4-b94a-4d6f-86a5-d6879c1504d1" />

vanaigr commented:
> NB: this removes the double buffering introduced in https://github.com/neovim/neovim/pull/32619 which is fundamentally incompatible with this (nvim core is supposed to keep track of long ranges by itself, without requiring a callback reinvoke blitz). Need to adjust the priorities some other way to fix the same issue.

Would it work if we pass `skip_until_row` and `skip_until_col` as `range_end_row` and `range_end_col` into `add_mark()`?

bfredl commented:
> There are some cases where the same node can be both a container and be contained within other nodes, in which case the tree depth would determine which highlighting should be applied, e.g. for the comment node. If we set its priority lower, then this would break

But in this case **strings** are used as the node where injections might happen. Thus even by the _very_ simple rule I suggested, by applying it what it actually means, "string" would get its priority lowered, not comment. For this particular filetype-scenario. So it would still work.

But my point more generally is rather that the core decoration mechanism allows for `2*16 = 65536` distinct priority levels. A plugin supporting a weird templating scenario where language A is embedded inside language B embedded inside another copy of language A might want to use a modified copy for highlights for the two different A layers, where priorities have been systematically adjusted for the inner layer to be consistently higher (nvim core could even help with this perhaps. it would be a setup cost, which is strongly preferred over making the highlighting hot loop more costly).

We even discussed making the "internal" priority space for providers larger before (32-bits), which is still possible but I'd like to see an attempt of making a dent into the full 16-bit dynamic range (a normal color dynamic range is 8-10 bits!) before that. Or nesting a lua-layer copy of `kvec(DecorRangeSlot)` just to emulate priority values which has not yet been properly specified.

clason commented:
>  might want to use a modified copy for highlights for the two different A layers

That is (currently) not possible.

bfredl commented:
Hmm I think it is possible (as a third party plugin) but inconvenient: you need to register it as two separate languages "A1" and "A2" (which can be done as you can load a language under a different name). The task at hand is to find a less kludgy abstraction which has the same effect:p

AFAIK "priority" is a concept which nvim owns entirely, and not something which upstream tree-sitter-highlight or tree-house and so on and so on cares about. What we need to figure out is the right spots a multi-lang filetype like this can enter the necessary information (which can involve _maths_, not just specifying up to 65536 levels manually)

clason commented:
Agreed; you can likely make it work manually (i.e., bypassing Neovim's automatic language resolution and manipulating the symbol), and this is what it may take for template languages (in plugins).

You are also correct that `priority` is a Neovim-only metadata, so we are free to handle this (or not) however we wish. Two things to note:
1. over time, we have become more averse to deviating from "vanilla" tree-sitter queries (where these are canonical: injections and highlights) even if some early addition will have to stay;
2. Neovim is unique in having an extensive, distributed parser support (through nvim-treesitter), so we get requests for situations that other editors simply don't handle.

Personally, I also think that (explicit) priorities are the wrong tool for handling cross-language conflicts, since we can't control the combination of arbitrary queries from arbitrary sources. (I also want to see less reliance on it for _inter_-language conflicts as tree-sitter has a number of inherent levers for that already.)

So after thinking about it, I'd prefer at least investing whether this could be handled by implicit "injection layer" priorities, with a manual offset via metadata if necessary. (My first thought was that injection queries have a stronger claim to being upstream compatible, but now I believe that this is less intrusive since a) it involves fewer modifications, b) unknown metadata is just ignored so safe and c) we might be able to interest upstream in a good proposal -- they suffer from template languages, same as we.)

This might require quite a bit of churn on nvim-treesitter, but since people are already calling it "chaotic", I no longer care about that.

clason commented:
I also think that we should distinguish two issues here:
1. the minimal example from https://github.com/neovim/neovim/issues/31777, which should Just Work and is hence a bug to fix in the implementation (whichever way works best for us);
2. injection sandwiches for template languages (which require a proper design of parsers/queries, ideally upstream -- i.e., Not Our Problem).

bfredl committed:
fix(treesitter): use subpriorities for tree ordering

This partially reverts 0b8a72b73934d33a05e20c255298e88cd921df32,
that is unreverts 15e77a56b711102fdc123e15b3f37d49bc0b1df1

"priority" is an internal neovim concept which does not occur in shared
queries. Ideally a single priority space should eventually be enough
for our needs. But as we don't want to poke at the usages of
priorities right now in the wider ecosystem,
introduce the "subpriorities" so that treesitter code can distinguish
highlights of the same priorities with different tree nesting depth.

This mainly affects `injection.combined` as parent-tree nodes might appear
in the middle of child-tree nodes which otherwise is not possible.

justinmk commented on runtime/doc/api.txt:
```diff
@@ -3336,6 +3336,12 @@ nvim_set_decoration_provider({ns_id}, {opts})
                    included. >
                     ["range", winid, bufnr, begin_row, begin_col, end_row, end_col]
 <
+                   In addition to returning a boolean, it is also allowed to
+                   return a `skip_row, skip_col` pair of integers. This
```
Any reason not to require this return type, and drop support for boolean?

justinmk replied:
> might be more convenient than `return 2147483647`

Could `math.huge` work? 

Generally we favor simpler types/logic, especially for uncommon use-cases (this is a rather low-level interface).

Not a blocker.
